스프링 자동차 경주 게임

givxq opened the pull request:
스프링 자동차 경주 게임

givxq committed:
[feature] SpringRacingCar
- racing car 도메인 추가

givxq committed:
[feature] SpringRacingCar
- racing car Controller 추가

givxq committed:
[feature] SpringRacingCar
- racing car Repository  추가

givxq committed:
[feature] SpringRacingCar
- README 추가

givxq committed:
[feature] SpringRacingCar
- PlayResult Repository 추가

givxq committed:
[feature] SpringRacingCar
- racing car Controller 수정

givxq committed:
[feature] SpringRacingCar
- racing car Model/Domain 수정

givxq committed:
[feature] SpringRacingCar
- README 수정

seokhongkim commented on README.md:
```diff
@@ -1 +1,64 @@
+  - 1, 2번 요구사항이 잘 수행되었다면 해당 기능을 페이지에서 확인 가능
+  - localhost:8080으로 접속하면 자동차 경주 게임 페이지에 접근
+
+# DB연동하기
```
이 내용도 구현해야 할 기능 목록으로 관리하고, 체크리스트 문법을 사용해보셔도 괜찮지 않을까요? :)

givxq replied:
넵 수정해보겠습니다 :)

seokhongkim commented on README.md:
JPA, lombok 사용을 위해서 라이브러리 의존성을 추가하셨을 것으로 보이는데요.
해당 내용이 커밋되지 않아서 빌드가 실패하네요.
또한, 자바 버전과 gradle 버전에도 변경이 있는 것으로 보입니다.
관련 내용 확인 부탁드립니다. :)

givxq replied:
gradle 설정관련 commit 추가하였습니다 :)

seokhongkim commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,25 @@
+
+    @PostMapping("/plays")
+    public ResponseEntity<RacingResponse> racingGame(@RequestBody RacingRequest racingRequest) {
+        return ResponseEntity.ok(racingCarService.startRacing(racingRequest));
```
메서드 체이닝을 이용해서 가독성을 높여보면 어떨까요?
```suggestion
        RacingResponse racingResponse = racingCarService.startRacing(racingRequest);
        return ResponseEntity.ok().body(racingResponse);
```

seokhongkim commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,58 @@
+@Entity
+@Getter
+@EntityListeners(AuditingEntityListener.class)
+@Table(name = "Cars")
```
사내에서 사용하는 테이블명 네이밍 컨벤션에 따라서 작성하신 것인지 궁금합니다. :)
다른 컨벤션을 따라서 이름을 지으신 것이 아니라면, 테이블명의 단수/복수와 관련된 글을 찾아보셔도 좋겠습니다.

givxq replied:
사내 테이블명은 코드로 관리되고 있습니다.
테이블명의 단수/복수 관련된 글을 찾아보았는데,
javascript, mongoDB쪽은 복수로 생성하는걸 권장하고 java진영은 단수로 생성하는걸 권장하네요.
저도 java를 사용하고 있으니 단수로 변경하겠습니다.👍

seokhongkim commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,58 @@
+@EntityListeners(AuditingEntityListener.class)
+@Table(name = "Cars")
+public class Car {
+    @Id
```
차의 이름을 id 로 사용하면 여러 경기에서 동일한 이름이 반복적으로 들어오는 경우 재사용이 불가할 것 같아요.

givxq replied:
네 말씀하신 내용이 맞습니다.

다만 현재 미션진행시에는 car 객체를 재사용하는 요건이 없고, 요구사항 진행시에도 해당 기능은 필요하지 않았습니다.
개인적으로 이번 미션 진행시에는 console로 구현한 객체들을 jpa를 사용하도록 변경시 얼마나 최소한의 수정으로 가능한지 보고 싶었습니다.

car객체의 재사용 문제는 기존 console 프로그램에서부터 있던 문제라고 생각했고, 
다음에 진행할 2차 미션에서 리팩토링을 진행하며 jpa의 id class로 복합키를 구현하며 개선해보려 했습니다.

seokhongkim commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,58 @@
+
+    private int position;
+
+    @CreatedDate
```
`CreatedDate`, `LastModifiedDate`를 모두 사용하신 이유가 있을까요?
`createAt`은 어떤 데이터를 담고 있는 컬럼일까요?

givxq replied:
ID 관련하여 말씀드린 내용과 관련있는 부분입니다.

현재 car table은 새로운 데이터가 생성되기도 하고 기존생성된 데이터의 수정도 일어나는 상태입니다.
(2차 미션 리팩토링 진행시 수정할 예정)

그런 이유에서 생성/수정 처리가 일어날때 시간을 기록하는 용도 입니다.

사실 미션 내용과는 무관한 내용이고, 개인적으로 jpa를 도입해보며 해당기능을 사용해보고 싶어 추가한 부분입니다.

seokhongkim replied:
생성하는 일시를 저장하고 이후 수정할 때마다 수정 일시를 관리하는 필드라면 `@LastModifiedDate`만 사용하셔도 될 것 같습니다. :)

seokhongkim commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,58 @@
+        this.name = name;
+    }
+
+    public void setName(String name) {
```
1. getter는 롬복의 `@Getter`를 이용해서 만드시고, setter는 코드로 직접 작성하신 이유가 궁금합니다.
2. setter가 필요할까요?

givxq replied:
최초에는 get/set 모두 코드로 작성했습니다.
get/set 코드가 길어짐에 따라 개인적으로 가독성이 떨어지는 코드처럼 보였습니다.

get의 경우는 조회성 기능이라 롬북으로 일괄 적용하였고,
set의 경우는 사이드이펙트가 생길 수 있어 필요하다고 생각한 부분만 코드로 작성하였습니다.

name도 set으로 수정하지 않고, build의 형태로 사용하도록 객체를 변경하려고 생각해보았지만
그렇게 변경할 경우 console에서 사용하는 로직도 변경을 해야하는 상황으로 보였습니다.

개인적으로 현재 미션은 console의 로직 변경을 최소화하고 spring 로직을 추가하고 싶어 set으로 구현하였고,
2차 미션인 리팩토링 진행시 해당 부분을 변경해보려 하였습니다.

seokhongkim replied:
네, 다음 단계에서는 setter 를 제거하는 방향으로 리팩터링 해보시면 좋을 것 같습니다! :)

seokhongkim commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public void move() {
+        if (CarRandom.isMove())
```
스타일 가이드에 따라 중괄호를 사용해보면 어떨까요? [Braces](https://google.github.io/styleguide/javaguide.html#s4.1-braces)
> Braces are used with if, else, for, do and while statements, even when the body is empty or contains only a single statement.

givxq replied:
네 사실 스타일가이드를 맞춰서 코딩을 할지 고민했었으나, 개인적으로 java에서 중괄호를 많이 사용하는 것을 선호하지 않고 있습니다...

개인적으로 다른 언어(python, javascript, elixir 등)와 비교하여 java 언어의 가장 불편한 점 중에 하나가 괄호였습니다. 

google에서도 java scrpit 컨벤션에서는 if문의 경우 1줄 일때 예외적으로 괄호를 생략하는걸 허용하고 있고,

linux, gnu 등의 코딩 컨벤션에서도 if문 이후 괄호를 생략하는걸 허용 하고 있습니다.

미션에서는 java 구글 컨벤션을 따라서 개발하라고 했지만.. 개인적인 선호의 따라 이부분은 양해 부탁드립니다..

seokhongkim replied:
그렇군요. :)
말씀하신 것처럼 선호의 영역이므로, 다음 리뷰에서는 중괄호에 대한 내용은 말씀드리지 않겠습니다.

저는 개인적으로 중괄호를 사용하는 것이 아래와 같은 장점이 있다고 생각해서 권장하기는 합니다.
* if, for code block 의 범위를 한 눈에 볼 수 있습니다. 한 줄의 코드만 있다고 하더라도, 해당 코드만 블록에 속한다는 것을 바로 알 수 있습니다.
* 코드 변경이 발생하여 코드 블록 내에 줄이 추가되거나 삭제될 때, 중괄호는 변경이 없으므로 추가/삭제된 코드의 diff 를 보는 것이 용이합니다.
* 코드 변경이 발생할 때, 코드 블록이 명확하므로 장애가 발생할 위험을 낮출 수 있습니다. 예를 들어, 한 줄의 코드만 있는 상황에서 해당 코드를 삭제했을 때, 다음 줄이 임의로 if, for 코드 블록에 포함되어 실행되는 상황을 방지할 수 있습니다.

seokhongkim commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,58 @@
+        this.position = position;
+    }
+
+    public void move() {
```
`move`와 `moveForward`라는 이름만 봤을 때, 메서드를 사용하는 사람이 각 메서드의 기능이 어떻게 다른지 알 수 있는 방법이 있을까요? `moveForward`는 내부에서 칸 이동을 위해서만 사용된다면 private 접근 제어자를 사용해보시면 어떨까요?

givxq replied:
말씀하신 내용에 공감합니다.
과거 racing car를 개발할때 부터 들었던 고민인데, 아시겠지만 private 메소드는 테스트코드에서 접근이 불가합니다.

과거 미션 진행시 내부에서 사용하는 메소드를 모두 private 생성하여 개발해보니
미션에서 요구하는 테스트코드들을 작성하기 힘들어졌고, 리뷰어 분과 이야기하면서 테스트코드에 적응하는게 목표이니 일단 모두 public으로 만들어보는게 좋을것 같다는 피드백을 받았습니다.

그런 이유로 현재 대부분의 메소드가 public으로 작성되어 있는 상태이고, 이 부분을 private로 추가적인 변경이 필요한 상태로 보입니다.

seokhongkim replied:
테스트 코드를 작성하는 연습을 하시기 위해 private 접근 제어자를 public 으로 변경하셨군요.
이제 테스트 코드를 작성하는 것에 익숙해지셨다면, 접근 제어자의 변경 없이 테스트를 작성해보시면 좋겠습니다. :)

private 메서드의 테스트는 public 메서드를 테스트하는 과정에서 함께 테스트가 되어야 합니다.
테스트를 위해 접근 제어자를 변경해야 할 필요가 있다면, 이는 메서드나 클래스의 설계를 되돌아봐야 한다는 신호일 수 있습니다.

seokhongkim commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public void move() {
+        if (CarRandom.isMove())
```
자동차가 이동하는 조건이 `CarRandom`에 의해 결정되는데, 테스트에서 특정 조건일 때 자동차가 멈춰있거나 움직이는지 검증하려면 어떻게 해야할까요? 현재는 `CarRandom`에서 결정하는 이동 조건을 테스트에서 통제하기 어려운 것으로 보이네요!

givxq replied:
네 말씀하신 내용에 공감합니다.👍
move 메소드를 오버로딩하여 해결했습니다.

seokhongkim replied:
테스트를 위해 메서드를 오버로딩하는 것은 프로덕션 코드가 테스트로 인해 변경된 것이므로 지양해야 한다고 생각합니다.
이 경우에는, `CarMoveUtil`의 정적 메서드에 의존하지 말고 `move`의 인자로 `CarMoveUtil`을 전달받거나 `Car`의 생성자를 통해 주입받고 필드로 관리하는 방법을 고민해보셔도 좋겠습니다.
정적 메서드에 의존하는 것은 코드가 분리된 것처럼 보이나, 실제로는 강한 정적 결합을 가지고 있고 테스트가 어려운 구조를 만들 수 있습니다.

seokhongkim commented on src/main/java/racingcar/domain/CarRandom.java:
```diff
@@ -0,0 +1,18 @@
+package racingcar.domain;
+
+public class CarRandom {
```
이 클래스가 하는 역할을 잘 설명해줄 수 있는 이름은 무엇이 있을까요?

givxq replied:
CarMoveUtil로 이름 변경해보았습니다.

seokhongkim commented on src/main/java/racingcar/domain/CarRandom.java:
```diff
@@ -0,0 +1,18 @@
+
+public class CarRandom {
+
+    final static int RANDOM_PIVOT_NUMBER = 4;
```
이 상수도 클래스 내부에서만 사용한다면 private 접근 제어자를 사용해보시면 좋겠습니다.

givxq replied:
네 공감합니다.

다른 코멘트에서도 말씀드렸지만, 과거 미션 진행시 private를 사용하지 않고 public으로 개발해보며 테스트 코드에서 사용하도록 가이드를 받았었습니다.

수정하도록 하겠습니다.

seokhongkim commented on src/main/java/racingcar/domain/PlayResult.java:
```diff
@@ -0,0 +1,53 @@
+
+@Entity
+@Getter
+@NoArgsConstructor
```
생성자를 public 으로 만들어야 할 필요가 있을까요?

givxq replied:
테스트 코드중 

PlayResult findplayResult = playResultRepository.findById(1L).orElse(null);

수행하기위해 추가하였습니다.

seokhongkim commented on src/main/java/racingcar/domain/PlayResult.java:
```diff
@@ -0,0 +1,53 @@
+@Entity
+@Getter
+@NoArgsConstructor
+@EntityListeners(AuditingEntityListener.class)
```
Audit 기능을 사용했을 때 CreatedDate 와 LastModifiedDate 가 어떤 식으로 값이 설정되는지도 학습해보시면 좋을 것 같습니다.
하나의 트랜잭션 내에서 다른 엔티티 여러개를 생성하거나 수정할 때 각 엔티티의 생성일시, 수정일시가 어떻게 저장되는지도 보시면 좋을 것 같아요.
이 부분을 학습, 테스트 해보시면 언제 Audit 기능을 사용하고, 사용하지 않는 게 좋을지 결정하시는 데 도움이 되리라 생각합니다.

givxq replied:
CreatedDate, LastModifiedDate는 엔티티 변경이 아닌 실제 db 데이터가 변경될때 발생합니다.
말씀하신 내용 중 사용하지 않는게 좋은 케이스는 어떤 경우일까요..?

seokhongkim replied:
여러 데이터를 하나의 트랜잭션에서 관리해보시면 위 어노테이션을 설정한 필드의 시간이 어떻게 저장되는지 확인해보실 수 있을 것 같습니다. :)

현업에서는 데이터를 관리할 때 하나의 트랜잭션에서 발생하는 데이터의 생성/수정 일시는 동일하게 관리하는 것이 중요하다고 생각합니다.
데이터 양이 많아질 때 시간을 이용하여 파티셔닝을 할 수도 있고, 특정 시점을 기준으로 데이터를 새로운 테이블로 이관하는 작업을 진행할 수도 있으며, 데이터 분석을 위해 저장된 데이터를 주기적으로 별도의 데이터 저장소로 동기화하는 작업도 있을텐데요. 이때, 함께 관리되어야 하는 데이터의 일시가 다르게 저장되면 문제가 발생할 수 있습니다.

seokhongkim commented on src/main/java/racingcar/domain/PlayResult.java:
```diff
@@ -0,0 +1,53 @@
+@Getter
+@NoArgsConstructor
+@EntityListeners(AuditingEntityListener.class)
+@SequenceGenerator(name = "PLAY_RESULT_SEQ_GENERATOR", sequenceName = "PLAY_SEQ", initialValue = 1, allocationSize = 1)
```
아이디 생성 방법으로 `SequenceGenerator`를 사용하신 이유가 있을까요?

givxq replied:
사내에서 오라클을 사용하고 있어서 SequenceGenerator 으로 id 생성하는 방법을 시도해 봤습니다.

seokhongkim commented on src/main/java/racingcar/domain/PlayResult.java:
```diff
@@ -0,0 +1,53 @@
+@NoArgsConstructor
+@EntityListeners(AuditingEntityListener.class)
+@SequenceGenerator(name = "PLAY_RESULT_SEQ_GENERATOR", sequenceName = "PLAY_SEQ", initialValue = 1, allocationSize = 1)
+@Table(name = "PLAY_RESULT", uniqueConstraints = {
```
테이블명 포맷이 `Cars`와 다르네요! 어떤 포맷을 사용하시나요?
컬럼명도 사용하시는 포맷이 있는지 궁금합니다.

givxq replied:
제가 생각하는 테이블명은 간단합니다.

car 테이블에서 가져오는 데이터는 대부분 집합의 형태입니다. (복수)

PlayResult 테이블에서 가져오는 데이터는 대부분 한 게임에 대한 결과 입니다. (단수)

테이블 네이밍 룰에 대해서 좀 더 찾아보니 외국에서도 많은 논쟁이 있는 부분이네요.
(https://vyaskn.tripod.com/object_naming.htm#Tables 여기서는 테이블명을 복수로 만드는 방법을 선호합니다.)

seokhongkim replied:
이 부분은 조금 더 명확한 기준을 세워보시는 것도 좋겠습니다.

데이터를 조회하는 형태는 서비스가 성장함에 따라서 다르게 조회를 할 수도 있고, 개발을 하는 사람마다 판단하는 "대부분"이라는 기준이 다를 수 있다고 생가합니다.

seokhongkim commented on src/main/java/racingcar/model/RacingRequest.java:
```diff
@@ -0,0 +1,16 @@
+package racingcar.model;
```
`model`과 `domain`은 어떻게 다른가요?
`model`은 MVC 패턴의 `model`로 생각될 수 있으므로, 다른 패키지명을 사용해보는건 어떨까요?

givxq replied:
model/domain의 차이는 비즈니스 로직 여부의 차이입니다.

https://wonit.tistory.com/258
이 글을 참고하고 개발한건 아니지만, 기본적으로 이분과 비슷한 생각으로 패키지를 나눴습니다.
개인적으로 해당 내용에 추가적으로 entity는 비스니즈 로직을 포함하고 있는 domain 영역으로 생성했습니다.

[model]
- request : Header에 Data(Entity)가 결합되어 실제 통신에 필요한 정보를 담고 있는 Request 클래스
- response : Request가 서비스 로직에 의해 가공된 Data를 Header와 결합된 Response 클래스
- header : 요청에 필요한 Header 정보를 담고 있는 클래스

seokhongkim commented on src/main/java/racingcar/repository/CarRepository.java:
```diff
@@ -0,0 +1,18 @@
+import racingcar.domain.Car;
+
+public interface CarRepository extends JpaRepository<Car, String> {
+    // Car move(Car car);
```
사용하지 않는 코드는 삭제하시면 좋겠습니다. :)

seokhongkim commented on src/main/java/racingcar/repository/CarRepository.java:
```diff
@@ -0,0 +1,18 @@
+
+    Car save(Car car);
+
+    Optional<Car> findByName(String name);
```
조회성 메서드도 모두 사용하지 않는 것으로 보이네요!

givxq replied:
테스트 코드에서는 사용하고 있습니다.

1차 미션시에는 db연동 이외의 db 관련 요구사항이 없어서 실제 코드에서는 사용하지는 않고 있습니다.

2차 미션에서는 아마도 db관련된 기능이 더 필요할듯한데요, 

현재는 JPA를 테스트해보면서 data가 db에 잘 저장되는지 확인하기 위해 테스트코드에서 사용중입니다.

seokhongkim commented on src/main/java/racingcar/service/RacingCarMain.java:
```diff
@@ -0,0 +1,18 @@
+            List<String> carNames = RacingInputCarView.getCarNames();
+            int targetDistance = RacingInputCarView.getTargetDistance();
+
+            // RacingCarService racingCarService = new RacingCarService();
```
콘솔에서는 현재 자동차 게임이 실행되지 않는 것일까요?
콘솔에서도 실행할 수 있도록 만들어보시면 좋겠습니다.

givxq replied:
1차 미션의 목표는 다음과 같은데요,

- 자동차 경주 코드 가져오기
- 웹 요청/응답 구현하기
- DB 연동하기
- 웹 애플리케이션을 구현하기 위해 기존 콘솔 애플리케이션을 수정하지 않습니다. 필요하다면 콘솔 애플리케이션에서 코드를 가져와도 됩니다. 중복 제거를 위한 리팩터링은 2단계에서 진행합니다.

말씀해주신 내용은 미션 목표에는 없는 내용으로 보이는데.. 어떤 의미일까요?

seokhongkim replied:
콘솔 환경에서 실행되는 코드와 웹 환경에서 실행되는 코드가 달라도 되지만, 두 프로그램을 개별적으로는 실행할 수 있어야 한다는 의미입니다. :)
콘솔 환경에서 실행되는 코드를 그대로 복사해온 상태에서 웹을 위한 코드를 추가로 만드셨다면, 코드의 중복은 발생할 수 있으나, 두 프로그램이 모두 실행은 되어야 한다고 생각했습니다!

seokhongkim commented on src/main/java/racingcar/service/RacingCarMain.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.List;
+import racingcar.view.RacingInputCarView;
+
+public class RacingCarMain {
```
웹의 메인은 `java.racingcar` 패키지에 있고, 콘솔의 메인은 `java.racingcar.service` 패키지에 있는 이유가 있을까요?

givxq replied:
1차 미션 내용 중

- 이전에 구현된 자동차 경주 미션 코드를 가져와서 racingcar 경로 하위에 위치시킵니다.

내용을 수행한 결과입니다.

seokhongkim commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -0,0 +1,104 @@
+import racingcar.view.RacingResultView;
+
+@Service
+@RequiredArgsConstructor
```
컨트롤러는 생성자를 직접 만드시고, 서비스는 `@RequireArgsConstructor`를 사용하신 이유가 있을까요?

givxq replied:
결과는 같겠지만 컨트롤러에서는 앞으로 소스를 수정하더라도 생성자의 변경이 없을것으로 예상되어 명시적으로 작성했습니다.

서비스에서는 앞으로 프로그램 개선에 따라 repository의 추가가 생길수도 있어 @RequireArgsConstructor 로 작성했습니다.

또 final 로 선언함에 따라 기존 console에 로직을 수정해야 하다보니,

1차 미션의 목표인 (웹 애플리케이션을 구현하기 위해 기존 콘솔 애플리케이션을 수정하지 않습니다.) 를 위반하게 되어

final을 제거하고 @Autowired 로 다시 수정하였습니다.

seokhongkim replied:
이 부분도 진범님의 구체적인 기준을 세워보시면 좋겠습니다. :)

모든 클래스에 `@RequiredArgsConstructor` 어노테이션을 이용해서 생성자 주입을 이용하는 방법도 있고, 필드에 `@Autowired` 어노테이션을 이용해서 필드 주입을 이용하는 방법도 모두 동작합니다.
코드를 함께 개발해야하는 동료가 있을 때 컨벤션을 정하고 지키는 것은 코드를 작성할 때 고민을 덜 수 있도록 도와주며, 전체 코드의 일관성을 유지할 수 있게 도와줍니다. 스스로 장단점을 비교해서 구체적인 기준을 세우시면, 협업을 하는 과정에서 컨벤션을 정할 때도 도움이 될 것 같습니다.

저는 개인적으로 `@RequiredArgusConstructor`, `@AllArgsConstructor` 사용을 지양합니다.
필드가 많아지는 경우 어노테이션을 통해 생성한 생성자는 필드의 순서 변경에 영향을 받아 생성자의 시그니처가 변하는데, 이 생성자를 사용하는 코드는 바로 반영되지 않는 경우가 있어서 리팩터링을 하는 과정에서 장애가 발생할 수 있습니다. 도메인이나 dto의 코드를 수정하는 과정에서 발생하는 문제가 많지만, 컨트롤러와 서비스에서 위 어노테이션을 사용하다보면 무의식중에 도메인이나 dto에서도 사용할 수 있다고 생각합니다.
또한 아예 사용을 금지함으로써 코드에 위 어노테이션이 포함되지 않도록 정적 분석이 가능하다는 장점이 있습니다.

seokhongkim commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -0,0 +1,104 @@
+    private final PlayResultRepository playResultRepository;
+
+    @Transactional
+    public RacingResponse startRacing(RacingRequest racingRequest) {
```
자동차 경주, 로또 미션에서 객체 지향에 대해 학습하셨으니 이를 바탕으로 웹의 자동차 경주도 객체를 이용해서 구현해보시면 어떨까요?
`자동차 게임`, `자동차 목록`을 객체로 만들어볼 수 있지 않을까요?
게임을 진행하고, 우승한 자동차 목록을 가져오는 책임을 별도의 클래스에게 위임해보시면 좋겠습니다.

위와 같이 변경한다면 서비스가 필요할까요? 서비스의 역할은 무엇일까요?

givxq replied:
말씀하신 내용은 공감합니다.👍

다만 이번 1차미션의 목표는 다음과 같은데요,

- 자동차 경주 코드 가져오기
- 웹 요청/응답 구현하기
- DB 연동하기

자동자 경주의 경우는 최초 진행한 미션이다 보니 일급객체를 사용하지 않은 코드였고,
현재 그 코드를 가져와 스프링을 연동하는 미션을 진행하다 보니 수정할 내용이 많이 보이는 상태인듯 합니다.

개인적으로 2차 미션을 JPA로 진행하게 되면 car 객체부터 설계가 달라지게 되어 많은 부분이 변경될듯 합니다.

또한 2차 미션은 리팩토링+기존 기능 수정 등의 목표가 포함되어 있으니 객체 설계를 변동하고, 프로그램의 구조를 바꾸는 내용들을 2차미션때 진행하려고 개인적으로 목표설정을 하였습니다.

seokhongkim commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -0,0 +1,104 @@
+
+    public List<String> getWinnerCars(List<Car> cars, int maxDistance) {
+        return cars.stream()
+                .filter(car -> car.getPosition() == maxDistance)
```
자동차의 위치를 꺼내서 비교하지 말고, 자동차에게 해당 위치에 있는지 물어보는 메서드를 구현해보면 어떨까요?

givxq replied:
네 말씀하신 내용에 공감합니다.

다만 1차미션에서의 car 객체의 구조를 변경하여야 할 필요가 있고, 

또한 이 부분에 대한 수정은 2차 미션의 목표인 

- 기존 기능 수정 - 출력 방석 수정
- 리팩터링 - 중복 코드 제거

에 가깝다고 생각하여 그때 말씀하신 내용을 적용해봐도 될까요?

seokhongkim commented on src/main/java/racingcar/view/RacingInputCarView.java:
```diff
@@ -0,0 +1,29 @@
+
+    public static List<String> getCarNames() {
+        System.out.println("경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).");
+        Scanner scan = new Scanner(System.in);
```
1. 이름은 가급적 축약하지 않고 작성해주시는 것이 좋습니다.
이름만 봤을 때 해당 객체가 무슨 타입인지 바로 알 수 있어서 코드를 읽을 때 가독성이 좋아진다는 장점이 있습니다. :)

2. `Scanner`를 필드로 관리해보면 어떨까요? 이름을 받을 때와 횟수를 받을 때 매번 생성해야 할 이유가 있을까요?

givxq replied:
말씀해주신 내용에 공감합니다.
개인적으로는 scan이라는 이름도 가독성은 좋다고 생각하여 이렇게 작성해보았네요 :)

Scanner 변수를 인스턴스 변수로 생성하는걸 말씀하시는걸까요?
그렇게 변경할 경우 재사용성은 좋아질수 있겠지만 굳이 heap 영역으로 변수를 다루고 싶지 않아 stack 영역의 변수로 사용했습니다.

seokhongkim replied:
말씀하신 것처럼 줄여 쓴 이름도 충분히 가독성은 좋습니다.
다만, 코드의 양이 많아지면 리팩터링을 하는 과정에서 코드를 검색하는 기능도 자주 활용하실텐데요. 이때, 줄여쓰기를 구체적인 규칙에 따라서 정해진 형태로 하지 않는다면, 어떤 형식으로 줄여썼는지 모르므로 검색에 소모되는 시간도 길어질 수 있습니다.
이 클래스 한정으로 생각했을 때는 줄여쓴 이름도 괜찮지만, 학습이므로 규칙을 만들고 연습을 하는 관점에서는 줄여쓰지 않는 것도 좋지 않을까 생각합니다. :)

seokhongkim commented on src/main/java/racingcar/view/RacingCarView.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Scanner;
+import java.util.stream.Collectors;
+
+public class RacingCarView {
```
사용하지 않는 클래스로 보이네요! 삭제해보시면 좋겠습니다. :)

givxq replied:
미션 진행내용을 들었을때 이전 소스를 그대로 가져와서 그 위에 spring 소스를 개발해보는걸로 들었습니다.

console 소스들을 지우는 내용은 2차 미션에 적혀있어서 1차미션때는 수정을 안했는데요,
말씀하신대로 수정해보겠습니다.

seokhongkim commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,22 @@
+    url: jdbc:h2:mem:testdb #H2 접속 정보
+    username: sa # H2 접속 시 입력할 username 정보 (원하는 것으로 입력)
+    password: # H2 접속 시 입력할 password 정보 (원하는 것으로 입력)
+  jps:
```
`jpa`의 오타일까요?

givxq replied:
앗 감사합니다👍

seokhongkim commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,22 @@
+      ddl-auto: create
+  output:
+    ansi:
+      enable: always
```
`enabled`의 오타일까요?

givxq replied:
감사합니다 👍

seokhongkim commented on src/main/resources/data.sql:
```diff
@@ -1,7 +1,8 @@
+--     created_at  DATETIME    NOT NULL default current_timestamp,
+--     PRIMARY KEY (id)
+-- );
+select * from dual;
```
이 쿼리는 왜 필요할까요?

givxq replied:
이 파일은 기본으로 미션을 진행하기 위해 공통으로 만들어주신 코드인데요,

저는 jpa로 개발하고 있고, 따로 테이블을 만들기 떄문에 필요하지 않은 파일입니다.

삭제하도록 하겠습니다.

seokhongkim commented on src/main/resources/data.sql:
```diff
@@ -1,7 +1,8 @@
+--     created_at  DATETIME    NOT NULL default current_timestamp,
+--     PRIMARY KEY (id)
+-- );
+select * from dual;
```
파일 마지막에 newline이 없다고 깃헙에서 경고를 띄워주네요.
파일 마지막의 newline은 왜 필요할까요? 학습해보신 후 IDE가 자동으로 파일 마지막의 newline을 추가하도록 설정해보셔도 좋겠습니다. :)

givxq replied:
이 파일은 기본으로 미션을 진행하기 위해 공통으로 만들어주신 코드인데요,

저는 jpa로 개발하고 있고, 따로 테이블을 만들기 떄문에 필요하지 않은 파일입니다.

삭제하도록 하겠습니다.

seokhongkim commented on src/test/java/racingcar/RacingCarApplicationTests.java:
```diff
@@ -6,8 +6,8 @@
 @SpringBootTest
 class RacingCarApplicationTests {
```
도메인 로직에 대한 테스트를 작성해보시면 좋겠습니다.

seokhongkim commented on src/test/java/racingcar/repository/CarRepositoryTest.java:
```diff
@@ -0,0 +1,35 @@
+import racingcar.domain.Car;
+
+@SpringBootTest
+@Rollback(false)
```
롤백하지 않도록 설정하신 이유가 있을까요?

givxq replied:
test 결과를 db에서도 확인해보려 사용하였습니다.

seokhongkim commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,58 @@
+@Getter
+@EntityListeners(AuditingEntityListener.class)
+@Table(name = "Cars")
+public class Car {
```
JPA를 사용할 때 equals, hashcode 는 무엇을 기준으로 정의해야 할까요?

givxq replied:
equlas, hashcode 는 테이블의 pk로 설정해야합니다.
현재 미션에서는 객체비교의 필요성이 없어보여 구현은 하지 않은 상태였었는데요, 추가해보도록 하겠습니다.👍

seokhongkim commented on src/main/java/racingcar/domain/PlayResult.java:
```diff
@@ -0,0 +1,53 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@Entity
```
어노테이션에 대해 말씀해주셔서 제 의견을 말씀드리기 전에, 잘 정리된 [블로그 글](https://betterprogramming.pub/the-problems-of-annotation-driven-development-9308491ef27a)이 있어서 공유 드립니다.

저는 개인적으로, JPA와 스프링을 팀에서 사용한다면 팀원들이 학습했고 이해할 수 있는 선에서 적절한 어노테이션 사용을 통해 얻을 수 있는 이점이 더 크다고 생각합니다. 즉, 위 글에서 언급한 장점이 더 크다고 느껴집니다. :)

진범님께서 말씀하신 단점인 실제 소스가 잘 안보인다는 점을 포함하여 정확한 동작이 무엇인지 알지 못한다는 점에도 공감하나, 이 부분은 어떤 기술을 사용할 때 우리가 모든 것을 100% 이해하지는 못하는 경우와 비슷하다고 생각합니다. 예를 들어, 제 경우에는 DB를 사용할 때 쿼리 실행 계획의 대략적인 내용은 알고 있으나, 이에 대한 구현(예를 들면, 복잡한 조인 쿼리가 실행되는 과정에서 어떻게 메모리를 할당하고 조인을 진행하여 최적화가 가능한지, 데이터 양이 커질 때 어떻게 메모리와 디스크 공간을 사용하는지)을 완전히 알지는 못 하는 점 등을 예로 들수 있겠습니다.

seokhongkim reviewed (changes requested):
진범님 안녕하세요.
스프링 자동차 경주 미션을 함께 진행할 리뷰어 김석홍입니다. 만나서 반갑습니다. 😀 

코드를 보고 궁금한 점과 DM으로 문의주신 내용에 대해 코멘트로 남겼습니다. 코멘트 확인 부탁드립니다.
코멘트 보시고 궁금한 점은 깃헙 코멘트 또는 슬랙 DM으로 남겨주시면 확인하는대로 답변 드리겠습니다.

givxq committed:
[feature] SpringRacingCar
- READMEc- 체크리스트 수정

givxq committed:
[feature] SpringRacingCar
- 메서드 체이닝으로 가독성 증가

givxq committed:
[feature] SpringRacingCar
- Car 테이블 단수로 변경

givxq committed:
[feature] SpringRacingCar
- Car 객체 hashcode, equlas 메소드 추가

givxq committed:
[feature] SpringRacingCar
- gradle 설정 commit

givxq committed:
[feature] SpringRacingCar
- moveForward private로 수정

givxq committed:
[feature] SpringRacingCar
- move 메소드 오버라이딩

givxq committed:
[feature] SpringRacingCar
- CarRandom -> CarMoveUtil 클래스 이름 변경

givxq committed:
[feature] SpringRacingCar
- test 코드 rollback 옵션 제거

givxq committed:
[feature] SpringRacingCar
- data.sql 삭제

givxq committed:
[feature] SpringRacingCar
- application.yml 오타수정

givxq committed:
[feature] SpringRacingCar
- RacingCarView 클래스 삭제

givxq committed:
[feature] SpringRacingCar
- 사용하지 않는 메소드 제거

givxq committed:
[feature] SpringRacingCar
- 미션 요구사항을 위한 sercive 생성자 변경

- 웹 애플리케이션을 구현하기 위해 기존 콘솔 애플리케이션을 수정하지 않습니다.
- 필요하다면 콘솔 애플리케이션에서 코드를 가져와도 됩니다. 중복 제거를 위한 리팩터링은 2단계에서 진행합니다.

givxq committed:
[feature] SpringRacingCarToJPA
- Cars 객체 생성

givxq committed:
[feature] SpringRacingCarToJPA
- JPA / Service 변경

givxq committed:
[feature] SpringRacingCarToJPA
- 필요없는 class 삭제

givxq committed:
Merge pull request #1 from givxq/springRacingCarToJPA

jpa 개발 merge

givxq committed:
[feature] SpringRacingCar
- 미사용 로직 삭제

givxq committed:
[feature] SpringRacingCar
- console 출력 변경

givxq committed:
[feature] SpringRacingCar
- 중복로직 제거

givxq committed:
[feature] SpringRacingCar
- 미사용 import 제거

givxq committed:
[feature] SpringRacingCar
- 중복 소스 제거

seokhongkim commented on src/test/java/racingcar/RacingCarApplicationTests.java:
```diff
@@ -1,13 +1,121 @@
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
 
 @SpringBootTest
```
테스트가 모두 실패하는 것으로 보이네요!
2단계에서는 도메인에 대한 단위 테스트를 작성해보셔도 좋겠습니다. :)

givxq replied:
test 코드 수행시 spring과 연동된 CommandLineRunner 코드가 수행되면서 오류가 나네요.
profile 설정값으로 test 일 경우에는 실행되지 않도록 변경했습니다.

2단계에서는 단위 테스트를 더 작성해보도록 하겠습니다. 👍

seokhongkim reviewed (approved):
진범님, 1단계 미션 코멘트 잘 반영하셨습니다. 👍 
1단계 PR 머지하겠습니다.
2단계 미션에서 뵙겠습니다. 😀
